src/lavinmq/queue/stream_queue_message_store.cr

```diff
@@ -88,6 +97,103 @@ module LavinMQ
         {msg_offset, segment, pos}
       end
 
+      def last_offset_by_consumer_tag(consumer_tag)
+        if pos = @consumer_offset_positions[consumer_tag]
+          tx = @consumer_offsets.to_slice(pos, 8)
+          return IO::ByteFormat::SystemEndian.decode(Int64, tx)
+        end
+      rescue KeyError
+      end
+
+      private def consumer_offset_positions
+        positions = Hash(String, Int64).new
+        slice = @consumer_offsets.to_slice
+        return positions if slice.size.zero?
+        pos = 0
+
+        loop do
+          ctag_length = IO::ByteFormat::LittleEndian.decode(UInt8, slice[pos, pos + 1])
+          break if ctag_length == 0
+          pos += 1
+          ctag = String.new(slice[pos..pos + ctag_length - 1])
+          pos += ctag_length
+          positions[ctag] = pos
+          pos += 8
+          break if pos >= slice.size
+        end
+        @consumer_offsets.resize(pos) # resize mfile to remove any empty bytes
+        positions
+      end
+
+      def save_offset_by_consumer_tag(consumer_tag, new_offset)
+        pos = 0_i64
+        begin
+          if pos = @consumer_offset_positions[consumer_tag]
+            buf = uninitialized UInt8[8]
+            IO::ByteFormat::LittleEndian.encode(new_offset.as(Int64), buf.to_slice)
+            @consumer_offsets.write_at(pos, buf.to_slice)
```

Comment:
```suggestion
            @consumer_offsets.pos = pos
            @consumer_offsets.write_bytes new_offset
```
less copying, see https://github.com/crystal-lang/crystal/blob/0aa30372c/src/io/byte_format.cr#L128-L132

Reply:
or else encode directly into the mmap:
```suggestion
            IO::ByteFormat::SystemEndian.encode(new_offset, @consumer_offsets.to_slice(pos, 8))
```

but `MFile#to_slice` returns a read_only Bytes/Slice.

Reply:
I can encode directly into the mmap though, but that requires us to add a read_only flag here. Should be OK? 

```
  def to_slice(pos, size, read_only = true)
    raise ClosedError.new if @closed
    raise IO::EOFError.new if pos + size > @size
    Bytes.new(buffer + pos, size, read_only: read_only)
  end
```

Reply:
ah, yes, forgot, `MFile` is append only, the pos is only for reading, which is different from `File`. Use your solution.